src/lib/communication/DY.Lib.Communication.RequestResponse.Invariants.fst

```diff
@@ -0,0 +1,211 @@
+  (comm_layer_aead_tag, aead_crypto_predicate_communication_layer)
+
+val has_communication_layer_reqres_crypto_predicates:
+  {|cinvs:crypto_invariants|} ->
+  prop
+let has_communication_layer_reqres_crypto_predicates #cinvs =
+  // Fix for the get_label function in the model code
+  cinvs.usages == default_crypto_usages /\
+  has_pke_predicate pke_crypto_predicates_communication_layer_and_tag /\
+  has_sign_predicate sign_crypto_predicates_communication_layer_and_tag /\
+  has_aead_predicate aead_crypto_predicates_communication_layer_and_tag
+
+
+(*** State Predicates ***)
+
+#push-options "--ifuel 2 --z3rlimit 25"
+let state_predicates_communication_layer {|crypto_invariants|}: local_state_predicate communication_states = {
+  pred = (fun tr prin sess_id st -> 
+    match st with
+    | ClientSendRequest {server; payload; key} -> (
+      let client = prin in
+      is_knowable_by (comm_label client server) tr payload /\
+      is_secret (comm_label client server) tr key /\
+      key `has_usage tr` (AeadKey comm_layer_aead_tag empty)
+    )
+    | ServerReceiveRequest {payload; key} -> (
+      let server = prin in
+      is_knowable_by (principal_label server) tr key /\
+      is_knowable_by (get_label tr key) tr payload /\
+      (
+        key `has_usage tr` (AeadKey comm_layer_aead_tag empty) \/
+        is_publishable tr key
+      )
+    )
+    | ClientReceiveResponse {server; payload; key} -> (
+      let client = prin in
+      is_knowable_by (comm_label client server) tr payload /\
+      is_secret (comm_label client server) tr key
+    )
+  );
+  pred_later = (fun tr1 tr2 prin sess_id state -> ());
+  pred_knowable = (fun tr prin sess_id state -> ());
+}
+#pop-options
+
+val state_predicates_communication_layer_and_tag: 
+  {|crypto_invariants|} ->
+  dtuple2 string local_bytes_state_predicate
+let state_predicates_communication_layer_and_tag #cinvs =
+  (|local_state_communication_layer_session.tag, local_state_predicate_to_local_bytes_state_predicate (state_predicates_communication_layer #cinvs)|)
+
+val has_communication_layer_state_predicates:
+  {|protocol_invariants|} ->
+  prop
+let has_communication_layer_state_predicates #invs =
+  has_local_state_predicate state_predicates_communication_layer
+
+
+(*** Event Predicates ***)
+
+noeq
+type comm_reqres_higher_layer_event_preds (a:Type) {| parseable_serializeable bytes a |} = {
+  send_request: tr:trace -> client:principal -> server:principal -> payload:a -> key_label:label -> prop;
+  send_request_later:
+    tr1:trace -> tr2:trace ->
+    client:principal -> server:principal -> payload:a -> key_label:label ->
+    Lemma
+    (requires
+      send_request tr1 client server payload key_label /\
+      bytes_well_formed tr1 (serialize a payload) /\
+      tr1 <$ tr2
+    )
+    (ensures
+      send_request tr2 client server payload key_label
+    )
+  ;
+  send_response: tr:trace -> server:principal -> payload:a -> prop;
```

Comment:
As noted earlier, we may in the future want to talk about the request and response together at some point, in which case at least one of these predicates probably needs to have both request and response payloads as parameters? Maybe we'll find other useful parameters to give the higher layer access to as we start using this more as well.